lib/Dialect/TritonGPU/Transforms/FoldTensorSelectIntoIf.cpp

```diff
@@ -0,0 +1,101 @@
+#include "mlir/Transforms/Passes.h"
+#include "triton/Dialect/TritonGPU/IR/Dialect.h"
+#include "triton/Dialect/TritonGPU/Transforms/Passes.h"
+#include "triton/Dialect/TritonGPU/Transforms/Utility.h"
+
+#include <memory>
+
+using namespace mlir;
+using namespace triton;
+using namespace triton::gpu;
+
+// TODO(jlebar): These autogenerated headers (and the passes they declare)
+// should be included within a namespace, but we have to do it consistently.
+#define GEN_PASS_CLASSES
+#include "triton/Dialect/TritonGPU/Transforms/Passes.h.inc"
+
+class FoldTensorSelectIntoIfPass
+    : public TritonGPUFoldTensorSelectIntoIfBase<FoldTensorSelectIntoIfPass> {
+public:
+  explicit FoldTensorSelectIntoIfPass() = default;
+
+  void runOnOperation() override {
+    MLIRContext *context = &getContext();
+    ModuleOp m = getOperation();
+
+    // Go over the arith.select ops, look if there is an if
+    // with the same condition, and if so, fold the select into the if.
+    DenseMap<arith::SelectOp, scf::IfOp> selectToIf;
+    m.walk([&](arith::SelectOp selectOp) {
+      // Look if there is an if in the same block, with the same condition.
+      auto *parentBlock = selectOp->getBlock();
+      Value condition = selectOp.getOperand(0);
+      // Get condition's users
+      for (auto user : condition.getUsers()) {
+        if (auto ifOp = dyn_cast<scf::IfOp>(user)) {
+          if (ifOp->getBlock() == parentBlock) {
+            selectToIf[selectOp] = ifOp;
```

Comment:
I believe there are extra conditions for the transformation to be correct. The IfOp needs to dominate all users of the select and the IfOp needs to be dominated by the selectOp (or at least by all its operands.

Reply:
Good catch, thank you!